### DISCUSSION TITLE
Tracking animal for open field analysis

### DISCUSSION THREAD

**Discussion by marta-vp** [bonsai-rx/bonsai#1103] [Q&A]
Hello!
I am a new Bonsai user and I have always been a slow learner re software. I am trying to make a workflow to analyse videos of open field.

This is what I want to do:
- record the video of one mouse in the open field cage
- track the movement of the animal and record the trajectory in a new video or picture
- measure the distance done by the animal
- measure the time spent in the center of the open field cage
- measure the time spent in the outer region of the open field cage

I have checked the available tutorials and the related questions from other users, which has helped me to start a workflow. But, to be honest, I am not sure if I am doing it correctly, so a bit (or a lot!) of help would be much appreciated.

Besides trying to do all the points listed above, I want to understand the .csv that I get at the end, I do not know where I can check what each column means.

![image](https://user-images.githubusercontent.com/115236656/198256367-dd588173-a92f-4ccf-9cf3-1036bbe413c2.png)

[OpenField.zip](https://github.com/bonsai-rx/bonsai/files/9878280/OpenField.zip)

Thank you in advance!

Marta

**Comment by diegogallo96** [comment]
I have the same problem.. i cant rescale in the corret way

**Comment by RoboDoig** [comment]
Hi Marta,

Have a look at the attached workflow for a rough idea of how to achieve what you need. I used a different video for the example so you may need to adjust some settings for your case.

![image](https://user-images.githubusercontent.com/11314305/220735772-9bf90038-d308-4160-a7e3-19dcb042fab4.png)

Let me try and address some of your questions:

**.csv output**
The CsvWriter will write a csv file with columns in order of how they are passed in Bonsai, e.g. if you have a Centroid, Value1, Value2 being fed via a zip into the CsvWriter your data columns will be [Centroid.X, Centroid.Y, Value1, Value2]. There is an option the CsvWriter node to IncludeHeader which will preserve the name of each stream in the data file. One trick is to use an ExpressionTransform to explicitly name each part of the stream which you'll see in the attached workflow.

![image](https://user-images.githubusercontent.com/11314305/220736930-659f6ed3-4796-4ce1-88c0-b5e6a1a805ae.png)

**Time spend in different regions**
The RoiActivity and RoiActivityDetected nodes are useful for this. With RoiActivity you can specify multiple regions of the image to 'observe' for activity. RoiActivityDetected allows you to specify a threshold for when activity in the ROI is counted as 'detected'. I've added an example in the attached workflow. In your case, it's probably best to use these nodes branching from the HsvThreshold node, as this is a binarised form of your input video relating to mouse position.

**Distance and trajectory**
In your original workflow the velocity and acceleration calculations were inside SelectMany nodes. Usually a SelectMany is used to produce multiple observable streams from an input, e.g. for repeating a calculation several times in parallel. For your case I think it is sufficient to use a GroupWorkflow for the same task. I've added an example in the attached workflow.

I also noticed in your acceleration calculation that you used the Timestamp node to get to a dv/dt value. You should be a bit careful with this as the Timestamp node here represents when Bonsai processed the frame rather than the true frame time from the video. For example, if you increased the playback rate of the FileCapture, you would increase the acceleration values that are calculated, even though the actual mouse acceleration didn't change. It's better to try and extract 'true' frame capture times for this calculation - how you do this depends on the camera you use but an example would be to use an LED in frame that flashes with a defined interval during your behavior recordings.

To visualize the animal trajectory here I've added a LineGraph visualizer.

[OpenFieldRevision.zip](https://github.com/bonsai-rx/bonsai/files/10807677/OpenFieldRevision.zip)

**Reply by kiakiatjytjy** [reply]
Hi [RoboDoig],

That explanation was very helpful. I am also another new user using trying to use bonsai for analyzing (Tmaze videos), may I ask  if you might know how we could combine the binary data (for example timeframes from when the animal passed through ROI) to the timestamp data that was generated during file capture?
![Sample Analysis pipeline](https://user-images.githubusercontent.com/95610609/225560682-96a86bb0-d39c-4903-af88-9b6dc15497e1.PNG)

My generated .csv file seems to be showing 0 and 1 in one column but no other corresponding time/labels at the side
![image](https://user-images.githubusercontent.com/95610609/225562030-8099c630-2975-4e49-a251-72874b544310.png)
![image](https://user-images.githubusercontent.com/95610609/225562078-26ad43d8-e6b9-469e-8a71-0eed55d43499.png)


Does this 1,0 follow the number of frames? just like how the original 'video capture > timestamps' were generated based on the frames?
![image](https://user-images.githubusercontent.com/95610609/225561024-0bb36705-b187-422a-918d-3b4f6c2b4f9e.png)


Thanks and Looking forward to hearing from you!

**Reply by RoboDoig** [reply]
Hi @jyntn - can you send a .zip with your workflow? I will take a look and get back to you.

**Reply by kiakiatjytjy** [reply]
Hi @RoboDoig,

I figured that the output 0,1 should be aligned with the video timestamps generated during the workflow for Tmaze videos. 
However, when I checked, I realised that the number of lines are different for the behavioral analysis output (eg. 8371 lines) compared to the video timestamps (8396 lines). 
Might you have an idea of how I might be able to reconcile that?

Attached are the two workflows!
Thanks for taking a look!
[behavior analysis.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/11210676/behavior.analysis.bonsai.zip)
[analog and digital only_202208016.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/11210685/analog.and.digital.only_202208016.bonsai.zip)

**Reply by RoboDoig** [reply]
Hi @jyntn - as I understand it you would like to save a csv with the video frame data paired with the times that an ROI was entered. Why not try something like this?

![image](https://user-images.githubusercontent.com/11314305/231486436-b0a55514-847a-45af-ad73-723fd9ac554b.png)

The csvwriter will not save timestamps unless you explicitly pass time values to the writer. In the workflow above, you will generate as many csv rows as there are frames in your FileCapture, so the rows are numbered by frame index.

I'm not sure I understand your other question, which two outputs are you comparing to get the 8371 vs. 8396 lines?

**Reply by tnkeo** [reply]
Hello, is there a way to make the ROIs for the center vs. edge exact? Or if there is a way to subtract total time of the trial by the amount of time spent in just the center so that I don't need to have two ROIs. Sorry, I am very new to Bonsai.

**Comment by aaronjnorris** [comment]
Maybe this is helpful as an example? 
https://wustl.box.com/v/openfiledexample
https://groups.google.com/g/bonsai-users/c/OO6dxm6kKH8/m/TR2bBQQHAQAJ

**Comment by BaixiH** [comment]
Hi everyone, this is Alex. I also plan on using Bonsai to track animal motion (coordinates) and Python to calculate the percentage of time spent in the center zone and average velocity. However, there is trouble when the output data (Coordinates) are restricted to a certain range. Can anyone help me? 

[OPF.zip](https://github.com/user-attachments/files/17518798/OPF.zip)
